dotnet/src/Microsoft.Agents.AI.Hosting.OpenAI/Responses/Converters/AfItemResourceGenerator.cs

```diff
@@ -0,0 +1,306 @@
+
+using System;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Agents.AI.Hosting.OpenAI.Responses.Common;
+using Microsoft.Agents.AI.Hosting.OpenAI.Responses.Common.Id;
+using Microsoft.Agents.AI.Hosting.OpenAI.Responses.Invocation;
+using Microsoft.Agents.AI.Hosting.OpenAI.Responses.Invocation.Stream;
+using Microsoft.Agents.AI.Hosting.OpenAI.Responses.Models;
+using Microsoft.Extensions.AI;
+
+namespace Microsoft.Agents.AI.Hosting.OpenAI.Responses.Converters;
+
+/// <summary>
+/// Generates item resources from agent run response updates.
+/// </summary>
+internal sealed class AfItemResourceGenerator
+    : INestedStreamEventGenerator<IEnumerable<ItemResource>>
+{
+    /// <summary>
+    /// Gets or sets the sequence number generator.
+    /// </summary>
+    public required ISequenceNumber Seq { get; init; }
+
+    /// <summary>
+    /// Gets or sets the updates to process.
+    /// </summary>
+    public required IAsyncEnumerable<AgentRunResponseUpdate> Updates { get; init; }
+
+    /// <summary>
+    /// Gets the sequence number generator for groups.
+    /// </summary>
+    private ISequenceNumber GroupSeq { get; } = SequenceNumberFactory.Default;
+
+    /// <summary>
+    /// Gets the context in which the agent invocation is executed.
+    /// </summary>
+    public required AgentInvocationContext Context { get; init; }
+
+    /// <summary>
+    /// Gets or sets the action to be invoked when there is an update on usage.
+    /// </summary>
+    public Action<ResponseUsage>? NotifyOnUsageUpdate { get; init; }
+
+    /// <inheritdoc/>
+    public async IAsyncEnumerable<NestedEventsGroup<IEnumerable<ItemResource>>> GenerateAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        await foreach (var updateGroup in this.Updates.ChunkOnChangeAsync(this.IsChanged, cancellationToken: cancellationToken).ConfigureAwait(false))
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            yield return this.CreateGroup(updateGroup);
+        }
+    }
+
+    private bool IsChanged(AgentRunResponseUpdate? previous, AgentRunResponseUpdate? current) => previous != null && current != null && this.Changed(previous, current);
+
+    /// <summary>
+    /// Determines whether the message ID has changed between two agent run response updates.
+    /// </summary>
+    /// <param name="previous">The previous <see cref="AgentRunResponseUpdate"/> instance to compare.</param>
+    /// <param name="current">The current <see cref="AgentRunResponseUpdate"/> instance to compare.</param>
+    /// <returns><see langword="true"/> if the message ID of the current update differs from the previous update; otherwise, <see
+    /// langword="false"/>.</returns>
+    private bool Changed(AgentRunResponseUpdate previous, AgentRunResponseUpdate current)
```

Comment:
In ToAgentRunResponse / ToChatResponse, this partitioning happens based on MessageId, AuthorName, and Role (we also avoid introducing boundaries when those transition from null to non-null). We should probably ensure the semantics stay in sync. I wonder if there's another API the core libs could expose that this could just use.

Reply:
I will copy the logic directly and we can remove it later if we add a method to check for this form of equality, eg `AgentRunResponseUpdate.IsSameMessage(AgentRunResponseUpdate? first, AgentRunResponseUpdate? second)`